tests/scripts/thread-cert/border_router/nat64/test_with_infrastructure_prefix.py

```diff
@@ -0,0 +1,146 @@
+#     notice, this list of conditions and the following disclaimer.
+#  2. Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#  3. Neither the name of the copyright holder nor the
+#     names of its contributors may be used to endorse or promote products
+#     derived from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 'AS IS'
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+import unittest
+
+import thread_cert
+
+# Test description:
+#   This test verifies the advertisement of infrastructure NAT64 prefix in Thread network.
+#
+#
+# Topology:
+#
+#   ----------------(eth)--------------------
+#           |
+#          BR (with DNS64 on infrastructure interface)
+#           |
+#        ROUTER
+#
+
+BR = 1
+ROUTER = 2
+
+# The prefix is set smaller than the default infrastructure NAT64 prefix.
+SMALL_NAT64_PREFIX = "2000:0:0:1:0:0::/96"
+
+
+class Nat64SingleBorderRouter(thread_cert.TestCase):
+    USE_MESSAGE_FACTORY = False
+
+    TOPOLOGY = {
+        BR: {
+            'name': 'BR',
+            'allowlist': [ROUTER],
+            'is_otbr': True,
+            'version': '1.2',
+        },
+        ROUTER: {
+            'name': 'Router',
+            'allowlist': [BR],
+            'version': '1.2',
+        },
+    }
+
+    def test(self):
+        br = self.nodes[BR]
+        router = self.nodes[ROUTER]
+
+        br.start()
+        self.simulator.go(5)
+        self.assertEqual('leader', br.get_state())
+
+        router.start()
+        self.simulator.go(5)
+        self.assertEqual('router', router.get_state())
+
+        # Case 1 BR advertise the infrastructure prefix
+        infra_nat64_prefix = br.get_br_nat64_prefix()
+
+        self.assertEqual(len(br.get_netdata_nat64_prefix()), 1)
+        nat64_prefix = br.get_netdata_nat64_prefix()[0]
+        self.assertEqual(nat64_prefix, infra_nat64_prefix)
+
+        # Case 2 Withdraw infrastructure prefix when a smaller prefix in medium
+        # preference is present
+        br.add_route(SMALL_NAT64_PREFIX, stable=False, nat64=True, prf='med')
```

Comment:
It's a little strange that we are adding the small NAT64 prefix on the BR itself.
Such test usually use another BR for adding a smaller NAT64 prefix.

Reply:
It's not straight forward to set up another BR with a smaller infrastructure NAT64 prefix. So using this way to mock the scenario and verify that the withdraw logic works.